lib/s3.persist.stream.js

```diff
@@ -0,0 +1,118 @@
+var AWS = require('aws-sdk');
+var s3 = new AWS.S3();
+var s3Stream = require('s3-upload-stream')(s3);
+var _ = require('underscore');
+
+var stream = require('stream');
+var util = require('util');
+
+var rotateHistoryCount = 3;
+var PassThrough = stream.PassThrough;
+
+function S3PersistStream(options) {
+  if (!(this instanceof S3PersistStream)) {
+    return new S3PersistStream(options);
+  }
+  PassThrough.call(this, options);
+
+  if (!options.filename) {
+    throw new Error('Filename must be specified for S3 persist stream');
+  }
+
+  if (!options.organization) {
+    throw new Error('Organization must be specified for S3 persist stream');
+  }
+
+  if (!options.stream_id) {
+    throw new Error('Stream ID must be specified for S3 persist stream');
+  }
+
+  this._filename = util.format('%s/%s/%s', options.organization, options.stream_id, options.filename);
+  this._metadata_version = options.filename;
+  this._organization = options.organization;
+  this._stream_id = options.stream_id;
+
+
+  this.uploaded = null;
+  this._file = null;
+  this._writeError = null;
+  this._rotateError = null;
+  this._rotateMessage = null;
+}
+
+util.inherits(S3PersistStream, PassThrough);
+
+S3PersistStream.prototype._write = function (chunk, enc, done) {
+  var self = this;
+
+  if (!this._writeError && this._file == null) {
+    try {
+      this._file = s3Stream.upload({
+        "Bucket": "bp-objects-cache",
+        "Key": this._filename
+      });
+
+      this._file.on('error', function(e) {
+        self._writeError = e;
+      })
+
+      /*
+      this._file.on('part', function(details) {
+      });
+      */
+
+      this._file.on('uploaded', function(details) {
+        self.rotate();
+        self.emit('persistdone', details);
+      })
+
+    } catch (e) {
+      this._writeError = e;
+    }
+
+  }
+
+  self.push(chunk);
+
+  if (!this._writeError) {
+    return this._file.write(chunk, enc, done);
```

Comment:
This means that the backpressure is a function of the S3 upload rate and not the next stream in the metadata.